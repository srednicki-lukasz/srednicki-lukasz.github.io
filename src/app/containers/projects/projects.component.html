<div class="h-100 card rounded-0 rounded-start">
  <div class="card-body d-flex align-items-center justify-content-center p-0">
    @if (isLoading()) {
    <div class="spinner-border text-light" role="status">
      <span class="sr-only"></span>
    </div>
    } @else {
    <div class="h-100 w-100 scrollable-container">
      <table class="table table-dark border-muted">
        <tbody>
          <tr class="portfolio-repository text-start">
            <td>
              <p class="p-0 m-0">{{ portfolioRepository().name }}</p>
              <small style="font-size: 0.8rem" class="accent">>_ {{ portfolioRepository().description || '—' }}</small>
            </td>
            <td class="align-middle text-center pe-0" style="width: 70px">
              <a [href]="portfolioRepository().html_url" target="_blank" class="btn text-light" title="GitHub">
                <span class="bi-github"></span>
                <span class="bi-chevron-right small ms-2"></span>
              </a>
            </td>
          </tr>

          @for (repository of repositoriesWithoutPortfolioRepository(); track repository.id) {
          <tr class="text-start">
            <td>
              <p class="p-0 m-0">{{ repository.name }}</p>
              <small style="font-size: 0.8rem" class="accent">>_ {{ repository.description || '—' }}</small>
            </td>
            <td class="align-middle text-center pe-0" style="width: 70px">
              <a [href]="repository.html_url" target="_blank" class="btn text-light" title="GitHub">
                <span class="bi-github"></span>
                <span class="bi-chevron-right small ms-2"></span>
              </a>
            </td>
          </tr>
          }
        </tbody>
      </table>
    </div>
    }
  </div>
</div>
