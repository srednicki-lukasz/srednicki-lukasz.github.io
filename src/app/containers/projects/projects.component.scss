@import '../../../styles.scss';

:host {
  @include router-base;
  padding: 12px 0 12px 12px;
}

.card {
  color: var(--color);
  overflow: hidden;
  box-shadow: var(--card-box-shadow);
  background-color: var(--card-background);
  background-image: url('../../../../public/texture.svg');
  background-size: cover;
  background-repeat: no-repeat;
  background-position: center top;

  .card-body {
    overflow: hidden;

    .scrollable-container {
      overflow: hidden;
      overflow-y: auto;

      .table {
        backdrop-filter: blur(4px);
        background-color: #21252952;

        .portfolio-repository {
          background: rgba($color: #1b9c85, $alpha: 0.1);
        }

        td {
          padding: 8px 16px;
          background: transparent;
        }
      }
    }
  }
}

.btn {
  display: inline-flex;
  overflow: hidden;
  position: relative;
  align-items: center;

  .bi-chevron-right {
    opacity: 0;
    display: inline-block;
    transform: translateX(-50%);
    transition: transform 0.3s ease;
  }

  &:hover .bi-chevron-right {
    opacity: 1;
    transform: translateX(0);
  }
}
