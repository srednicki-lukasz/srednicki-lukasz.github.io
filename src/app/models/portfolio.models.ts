import { components } from '@octokit/openapi-types';

export type Repository = components['schemas']['repository'];
