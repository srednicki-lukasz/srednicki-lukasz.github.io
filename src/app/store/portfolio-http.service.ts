import { HttpClient } from '@angular/common/http';
import { inject, Injectable } from '@angular/core';
import { firstValueFrom } from 'rxjs';
import { Repository } from '../models/portfolio.models';

@Injectable()
export class PortfolioHttpService {
  private readonly httpClient = inject(HttpClient);
  private readonly origin = 'https://api.github.com';

  public fetchRepositories(): Promise<Repository[]> {
    return firstValueFrom(this.httpClient.get<Repository[]>(`${this.origin}/users/srednicki-lukasz/repos`));
  }
}
