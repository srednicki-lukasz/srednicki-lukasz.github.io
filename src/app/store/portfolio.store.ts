import { computed, inject } from '@angular/core';
import { patchState, signalStore, withComputed, withMethods, withState } from '@ngrx/signals';
import { Repository } from '../models/portfolio.models';
import { PortfolioHttpService } from './portfolio-http.service';

const configRepositoryId = 347375901;
const portfolioRepositoryId = 978832136;

type PortfolioState = {
  isLoading: boolean;
  repositories: Repository[];
};

const initialState: PortfolioState = {
  isLoading: false,
  repositories: [],
};

export const PortfolioStore = signalStore(
  withState(initialState),
  withComputed(({ repositories }) => ({
    portfolioRepository: computed((): Repository => {
      return repositories().find(({ id }) => id === portfolioRepositoryId)!;
    }),
    repositoriesWithoutPortfolioRepository: computed((): Repository[] => {
      return repositories().filter(({ id }) => id !== configRepositoryId && id !== portfolioRepositoryId);
    }),
  })),
  withMethods((store, portfolioHttpService = inject(PortfolioHttpService)) => ({
    async fetchRepositories(): Promise<void> {
      if (store.repositories().length) return;
      patchState(store, { isLoading: true });

      const repositories = await portfolioHttpService.fetchRepositories();
      patchState(store, { isLoading: false, repositories });
    },
  }))
);
