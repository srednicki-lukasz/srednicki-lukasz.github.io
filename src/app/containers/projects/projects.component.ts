import { ChangeDetectionStrategy, Component, computed, inject, OnInit } from '@angular/core';
import { Repository } from '../../models/portfolio.models';
import { PortfolioStore } from '../../store/portfolio.store';

@Component({
  selector: 'projects',
  templateUrl: './projects.component.html',
  styleUrl: './projects.component.scss',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class ProjectsComponent implements OnInit {
  private readonly store = inject(PortfolioStore);

  isLoading = computed(this.computeIsLoading.bind(this));
  portfolioRepository = computed(this.computePortfolioRepository.bind(this));
  repositoriesWithoutPortfolioRepository = computed(this.computeRepositoriesWithoutPortfolioRepository.bind(this));

  public ngOnInit(): void {
    this.store.fetchRepositories();
  }

  private computeIsLoading(store = this.store): boolean {
    return store.isLoading();
  }

  private computePortfolioRepository(store = this.store): Repository {
    return store.portfolioRepository();
  }

  private computeRepositoriesWithoutPortfolioRepository(store = this.store): Repository[] {
    return store.repositoriesWithoutPortfolioRepository();
  }
}
