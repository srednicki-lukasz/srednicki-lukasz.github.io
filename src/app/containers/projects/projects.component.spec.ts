import { provideHttpClient } from '@angular/common/http';
import { importProvidersFrom } from '@angular/core';
import { RouterTestingModule } from '@angular/router/testing';
import { createComponentFactory, Spectator } from '@ngneat/spectator/jest';
import { Repository } from '../../models/portfolio.models';
import { PortfolioHttpService } from '../../store/portfolio-http.service';
import { PortfolioStore } from '../../store/portfolio.store';
import { createTestRepository } from '../../test/portfolio.test-data';
import { ProjectsComponent } from './projects.component';

const testRepositories = [createTestRepository(), createTestRepository(1)];

describe('ProjectsComponent', () => {
  let spectator: Spectator<ProjectsComponent>;

  let portfolioHttpService: jest.Mocked<PortfolioHttpService>;
  let resolveFetchRepositories: (repositories: Repository[]) => void;

  const createComponent = createComponentFactory({
    component: ProjectsComponent,
    providers: [
      PortfolioStore,
      PortfolioHttpService,
      provideHttpClient(),
      importProvidersFrom(RouterTestingModule),
      {
        provide: PortfolioHttpService,
        useValue: {
          fetchRepositories: jest
            .fn()
            .mockImplementation(() => new Promise<Repository[]>(resolve => (resolveFetchRepositories = resolve))),
        },
      },
    ],
    detectChanges: false,
  });

  beforeEach(() => {
    spectator = createComponent();
    portfolioHttpService = spectator.inject(PortfolioHttpService) as any;
  });

  it('should fetch repositories and match snapshot', async () => {
    expect(spectator.component.isLoading()).toBeFalsy();

    spectator.detectChanges();

    expect(portfolioHttpService.fetchRepositories).toHaveBeenCalled();
    expect(spectator.component.isLoading()).toBeTruthy();

    spectator.detectChanges();

    resolveFetchRepositories(testRepositories);
    await Promise.resolve();

    expect(spectator.component.isLoading()).toBeFalsy();
    expect(spectator.component.portfolioRepository()).toEqual(testRepositories[0]);
    expect(spectator.component.repositoriesWithoutPortfolioRepository()).toEqual([testRepositories[1]]);

    spectator.detectChanges();

    expect(spectator.fixture).toMatchSnapshot();
  });
});
