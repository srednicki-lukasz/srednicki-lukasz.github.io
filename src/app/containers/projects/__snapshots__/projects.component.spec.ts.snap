// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectsComponent should fetch repositories and match snapshot 1`] = `
<projects
  isLoading={[Function Function]}
  portfolioRepository={[Function Function]}
  repositoriesWithoutPortfolioRepository={[Function Function]}
  store={[Function _SignalStore]}
>
  <div
    class="h-100 card rounded-0 rounded-start"
  >
    <div
      class="card-body d-flex align-items-center justify-content-center p-0"
    >
      <div
        class="h-100 w-100 scrollable-container"
      >
        <table
          class="table table-dark border-muted"
        >
          <tbody>
            <tr
              class="portfolio-repository text-start"
            >
              <td>
                <p
                  class="p-0 m-0"
                >
                  srednicki-lukasz.github.io
                </p>
                <small
                  class="accent"
                  style="font-size: 0.8rem;"
                >
                  &gt;_ Portfolio App.
                </small>
              </td>
              <td
                class="align-middle text-center pe-0"
                style="width: 70px;"
              >
                <a
                  class="btn text-light"
                  href="https://github.com/srednicki-lukasz/srednicki-lukasz.github.io"
                  target="_blank"
                  title="GitHub"
                >
                  <span
                    class="bi-github"
                  />
                  <span
                    class="bi-chevron-right small ms-2"
                  />
                </a>
              </td>
            </tr>
            <tr
              class="text-start"
            >
              <td>
                <p
                  class="p-0 m-0"
                >
                  srednicki-lukasz.github.io
                </p>
                <small
                  class="accent"
                  style="font-size: 0.8rem;"
                >
                  &gt;_ Portfolio App.
                </small>
              </td>
              <td
                class="align-middle text-center pe-0"
                style="width: 70px;"
              >
                <a
                  class="btn text-light"
                  href="https://github.com/srednicki-lukasz/srednicki-lukasz.github.io"
                  target="_blank"
                  title="GitHub"
                >
                  <span
                    class="bi-github"
                  />
                  <span
                    class="bi-chevron-right small ms-2"
                  />
                </a>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</projects>
`;
